protocol/op-stack-manager-arch.md

```diff
@@ -0,0 +1,292 @@
+
+The current L2 chain deployment approach originates from a time with Hardhat, single L1 target, and a single monolithic set of features.
+Since then the system has migrated to Foundry and extended for more features, but remains centered around a single monolithic deploy-config for all its features.
+
+The interop team needs a way to configure new multi-L2 deployments: The number of ways to compose L2s in tests grows past what a single legacy config template can support.
+
+Outside of interop, deployment also seems increasingly complex and opaque, while it does not have to be, due to the same configuration and composability troubles.
+
+Part of the solution to this is described in https://github.com/ethereum-optimism/design-docs/pull/52,and the other part involves restructuring our Solidity config processing and deploy scripts.
+This design doc focuses on the L1 contract configuration and deployment part of the solution.
+
+# Alternatives Considered
+
+The primary alternative is for the interop team to not use OPSM for devnet and testnet deployments.
+This is not ideal for various reasons:
+
+- Duplication of effort: The interop team would need to develop and maintain their own deployment scripts, even though new ones will already be written for OPSM.
+- Divergence: With duplication, it's likely the interop team's deployment scripts diverge from the production deployment scripts, leading to inconsistencies and bugs.
+
+# Proposed Solution
+
+There are two steps to deploying L1 contracts:
+
+1. **Deploy Superchain Contracts**. This only occurs occasionally, as the Superchain contracts are shared between many OP chains.
+2. **Deploy Shared Implementation Contracts**. This occurs once per [contracts release](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/VERSIONING.md).
+3. **Deploy OP Chain Contracts**. This occurs for once every OP chain deployment.
+
+For each step we define the step inputs (configuration), sub-steps that occur, and the step outputs (artifacts).
+For each we start by listing the sub-steps, then define the inputs and outputs for each sub-step, and use those to define the inputs and outputs for the step as a whole.
+
+Sample input and output TOML files are included for each step.
+The exact structure of these files is too low-level to be in-scope for this design doc, so just consider them as examples.
+However, feel free to comment on specifics if you think they're important.
+
+TOML is used because it is what the superchain-registry is standardizing on, it can be parsed by forge and go, it supports comments (unlike JSON), and it does not have the [quirks](https://noyaml.com/) of YAML.
+
+Each step is a separate forge script that takes the input TOML file as the only input, and outputs a TOML file.
+To make the artifact self-contained, the outputs also include the inputs.
+This allows a single artifact to be used to convey all information needed about a deployment.
```

Comment:
Like config inputs, using an interface for outputs would also be nice. I.e. an interface of `output(<metadata>)`, and then a contract implementation that does the JSON state dump and any post-processing.

Reply:
Perhaps the forge scripts always postprocess to output a file like
```toml
[addresses]
# all output addresses here

[inputs]
# all inputs are nested under here
```

Since all outputs from these steps are just addresses

Reply:
same as https://github.com/ethereum-optimism/design-docs/pull/60#discussion_r1710274580 but for outputs